src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -342,6 +343,17 @@ public async ValueTask DisposeAsync()
     {
         if (Interlocked.Exchange(ref _state, StateDisposed) is not StateDisposed)
         {
+            // Complete outstanding subscriptions.
+            lock (_lock)
+            {
+                foreach (var item in _outgoingChannels)
+                {
+                    item.Writer.Complete();
+                }
+
+                _outgoingChannels = _outgoingChannels.Clear();
+            }
```

Comment:
This needs to be interlocked. The `_outgoingChannels` collection is not protected by any lock object.

```suggestion
            // Complete outstanding subscriptions.
            var channelsToComplete = Interlocked.Exchange(ref _outgoingChannels, ImmutableHashSet<Channel<ResourceViewModelChange>>.Empty);

            foreach (var channel in channelsToComplete)
            {
                channel.Writer.Complete();
            }
```

Reply:
I think without locking there is the chance of race conditions between subscribing and disposing. But disposing in the real world means the dashboard is being shutdown so it doesn't matter.

What purpose does lock have in this type?

Reply:
The lock in the subscribe method is needed to protect against collection modified exceptions when snapshotting the resources.